niff-1.02/niffsdk1.02/riffio/doc/man/chunks.man
.rn '' }`
''' $RCSfile$$Revision$$Date$
'''
''' $Log$
'''
.de Sh
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp
.if t .sp .5v
.if n .sp
..
.de Ip
.br
.ie \\n(.$>=3 .ne \\$3
.el .ne 3
.IP "\\$1" \\$2
..
.de Vb
.ft CW
.nf
.ne \\$1
..
.de Ve
.ft R

.fi
..
'''
'''
'''     Set up \*(-- to give an unbreakable dash;
'''     string Tr holds user defined translation string.
'''     Bell System Logo is used as a dummy character.
'''
.tr \(*W-|\(bv\*(Tr
.ie n \{\
.ds -- \(*W-
.ds PI pi
.if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\" diablo 12 pitch
.ds L" ""
.ds R" ""
.ds L' '
.ds R' '
'br\}
.el\{\
.ds -- \(em\|
.tr \*(Tr
.ds L" ``
.ds R" ''
.ds L' `
.ds R' '
.ds PI \(*p
'br\}
.\"	If the F register is turned on, we'll generate
.\"	index entries out stderr for the following things:
.\"		TH	Title 
.\"		SH	Header
.\"		Sh	Subsection 
.\"		Ip	Item
.\"		X<>	Xref  (embedded
.\"	Of course, you have to process the output yourself
.\"	in some meaninful fashion.
.if \nF \{
.de IX
.tm Index:\\$1\t\\n%\t"\\$2"
..
.nr % 0
.rr F
.\}
.TH chunks 3 "riffio" "14/Jun/96" "RIFFIO Documentation"
.IX Title "chunks 3"
.UC
.IX Name "chunks - create, navigate, and operate on chunks"
.if n .hy 0
.if n .na
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.de CQ          \" put $1 in typewriter font
.ft CW
'if n "\c
'if t \\&\\$1\c
'if n \\&\\$1\c
'if n \&"
\\&\\$2 \\$3 \\$4 \\$5 \\$6 \\$7
'.ft R
..
.\" @(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2
.	\" AM - accent mark definitions
.bd B 3
.	\" fudge factors for nroff and troff
.if n \{\
.	ds #H 0
.	ds #V .8m
.	ds #F .3m
.	ds #[ \f1
.	ds #] \fP
.\}
.if t \{\
.	ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.	ds #V .6m
.	ds #F 0
.	ds #[ \&
.	ds #] \&
.\}
.	\" simple accents for nroff and troff
.if n \{\
.	ds ' \&
.	ds ` \&
.	ds ^ \&
.	ds , \&
.	ds ~ ~
.	ds ? ?
.	ds ! !
.	ds /
.	ds q
.\}
.if t \{\
.	ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.	ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.	ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.	ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.	ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.	ds ? \s-2c\h'-\w'c'u*7/10'\u\h'\*(#H'\zi\d\s+2\h'\w'c'u*8/10'
.	ds ! \s-2\(or\s+2\h'-\w'\(or'u'\v'-.8m'.\v'.8m'
.	ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.	ds q o\h'-\w'o'u*8/10'\s-4\v'.4m'\z\(*i\v'-.4m'\s+4\h'\w'o'u*8/10'
.\}
.	\" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds v \\k:\h'-(\\n(.wu*9/10-\*(#H)'\v'-\*(#V'\*(#[\s-4v\s0\v'\*(#V'\h'|\\n:u'\*(#]
.ds _ \\k:\h'-(\\n(.wu*9/10-\*(#H+(\*(#F*2/3))'\v'-.4m'\z\(hy\v'.4m'\h'|\\n:u'
.ds . \\k:\h'-(\\n(.wu*8/10)'\v'\*(#V*4/10'\z.\v'-\*(#V*4/10'\h'|\\n:u'
.ds 3 \*(#[\v'.2m'\s-2\&3\s0\v'-.2m'\*(#]
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.ds oe o\h'-(\w'o'u*4/10)'e
.ds Oe O\h'-(\w'O'u*4/10)'E
.	\" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.	\" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.	ds : e
.	ds 8 ss
.	ds v \h'-1'\o'\(aa\(ga'
.	ds _ \h'-1'^
.	ds . \h'-1'.
.	ds 3 3
.	ds o a
.	ds d- d\h'-1'\(ga
.	ds D- D\h'-1'\(hy
.	ds th \o'bp'
.	ds Th \o'LP'
.	ds ae ae
.	ds Ae AE
.	ds oe oe
.	ds Oe OE
.\}
.rm #[ #] #H #V #F C
.SH "NAME"
.IX Header "NAME"
chunks \- create, navigate, and operate on chunks
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Sh "Chunk Creation"
.IX Subsection "Chunk Creation"
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkCreate()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkFinalize()\fR
.Sh "Chunk Navigation"
.IX Subsection "Chunk Navigation"
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkDescend()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkAscend()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkDataEnd()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkEnd()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkDataOffset()\fR
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkSeekData()\fR
.Sh "Chunk Operations"
.IX Subsection "Chunk Operations"
.Ip "\(bu" 4
.IX Item "\(bu"
\fIRIFFIOChunkIsList()\fR
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
.SH "RIFFIOChunkCreate"
.IX Header "RIFFIOChunkCreate"
Start a new chunk in a RIFF file.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     RIFFIOChunkCreate(RIFFIOFile *prf, RIFFIOChunk *pchunk) 
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
pchunk->fccId provides the new chunk id.
.Ip "\(bu" 4
.IX Item "\(bu"
pchunk->fccType must contain a type id for Lists and Forms and
is ignored otherwise.
.Ip "\(bu" 4
.IX Item "\(bu"
pchunk->sizeData does not need to be correct.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
Writes a chunk header to *prf according to *pchunk.
\fIRIFFIOChunkCreate()\fR will write pchunk->sizeData to the file.
.Ip "\(bu" 4
.IX Item "\(bu"
Leaves *prf positioned at the start of the new chunk's data,
.Ip "\(bu" 4
.IX Item "\(bu"
If *pchunk is a form, then *prf is marked with the proper
byte order.
.Sh "\s-1OBLIGATIONS\s0"
.IX Subsection "\s-1OBLIGATIONS\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
After writing the chunk contents, *pchunk
must be finished with \fIRIFFIOChunkFinalize()\fR.
.Ip "\(bu" 4
.IX Item "\(bu"
Don't change the <*pchunk> data members before calling
\fIRIFFIOChunkFinalize()\fR.
.Ip "\(bu" 4
.IX Item "\(bu"
If the pchunk->sizeData is correct when the chunk is finalized,
then \fIRIFFIOChunkFinalize()\fR will not update the chunk's size in
the file.
.Sh "\s-1ERRORS\s0"
.IX Subsection "\s-1ERRORS\s0"
On failure, the file position is undefined.
.SH "RIFFIOChunkFinalize"
.IX Header "RIFFIOChunkFinalize"
Finish writing a new chunk in a RIFF file.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     RIFFIOChunkFinalize(RIFFIOFile *prf, RIFFIOChunk *pchunk)
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
<*pchunk> was returned by RIFFIOChunkCreate
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
Updates the chunk size in the file and in <*pchunk> (if necessary).
.Ip "\(bu" 4
.IX Item "\(bu"
Writes a \s-1NUL\s0 pad byte (if necessary).
.Ip "\(bu" 4
.IX Item "\(bu"
Leaves the file positioned after the chunk.
.Sh "\s-1ERRORS\s0"
.IX Subsection "\s-1ERRORS\s0"
On failure, the file position is undefined.
.SH "RIFFIOChunkDescend"
.IX Header "RIFFIOChunkDescend"
Read a chunk header from a RIFF file.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     RIFFIOChunkDescend(RIFFIOFile *prf, RIFFIOChunk *pchunk)
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
<*prf> must be positioned at the start of a chunk.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
.Ip "\(bu" 4
.IX Item "\(bu"
Leaves the file positioned at the start of the chunk's contents
(either the first subchunk of a \s-1LIST\s0 or a normal chunk's data).
.Ip "\(bu" 4
.IX Item "\(bu"
If the chunk is a \s-1RIFF\s0 or \s-1RIFX\s0 form, then
marks the RIFFIOFile byte order accordingly.
.Ip "\(bu" 4
.IX Item "\(bu"
<*pchunk> holds the id, size, and possibly the
list type of the newly read chunk.
.Sh "\s-1ERRORS\s0"
.IX Subsection "\s-1ERRORS\s0"
On failure the file position is undefined.
.SH "RIFFIOChunkAscend"
.IX Header "RIFFIOChunkAscend"
Advance a RIFF file past a chunk.
.PP
.Vb 2
\&     RIFFIOSuccess
\&     RIFFIOChunkAscend(RIFFIOFile *prf, const RIFFIOChunk *pchunk)
.Ve
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
<*pchunk> was returned by either RIFFIOChunkDescend or
RIFFIOChunkFinalize.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
<*prf> is positioned at the end of <*pchunk>
.Sh "\s-1ERRORS\s0"
.IX Subsection "\s-1ERRORS\s0"
On failure, the file position is undefined.
.SH "RIFFIOChunkDataOffset"
.IX Header "RIFFIOChunkDataOffset"
Return the file position of a chunk's data (or list's first subchunk)
.PP
.Vb 2
\&     RIFFIOOffset
\&     RIFFIOChunkDataOffset(const RIFFIOChunk *pchunk)
.Ve
.SH "RIFFIOChunkSeekData"
.IX Header "RIFFIOChunkSeekData"
Seek to the start of a chunk's data or a list's contents
.PP
.Vb 2
\&     RIFFIOSuccess
\&     RIFFIOChunkSeekData(RIFFIOFile *prf, const RIFFIOChunk *pchunk)
.Ve
Note that this is different than \fIRIFFIOChunkDescend()\fR because
\fIRIFFIOChunkSeekData()\fR can be called regardless of the current
file postion.
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
<*prf> may be positioned anywhere.
.Sh "\s-1EXIT\s0"
.IX Subsection "\s-1EXIT\s0"
<*prf> is postioned at the start of a chunk's data or the first
subchunk of a list.
.Sh "\s-1ERRORS\s0"
.IX Subsection "\s-1ERRORS\s0"
On error, the file postion is undefined.
.SH "RIFFIOChunkIsList"
.IX Header "RIFFIOChunkIsList"
Return true if a chunk is a form or a list.
.PP
.Vb 2
\&     int
\&     RIFFIOChunkIsList(const RIFFIOChunk *pchunk)
.Ve
.SH "RIFFIOChunkDataEnd"
.IX Header "RIFFIOChunkDataEnd"
Return true if a file is positioned at the end of a chunk's data.
.PP
.Vb 2
\&     int
\&     RIFFIOChunkDataEnd(RIFFIOFile *prf, const RIFFIOChunk *pchunk)
.Ve
The chunk data's end may be before any NUL pad byte.
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
<*pchunk> must be a \*(L"clean\*(R" chunk, returned by either
\fIRIFFIOChunkDescend()\fR or \fIRIFFIOChunkFinalize()\fR.
.SH "RIFFIOChunkEnd"
.IX Header "RIFFIOChunkEnd"
Return true if a RIFF file is positioned at the end of a chunk.
.PP
.Vb 2
\&     int
\&     RIFFIOChunkEnd(RIFFIOFile *prf, const RIFFIOChunk *pchunk)
.Ve
The chunk end is after any NUL pad byte.
.Sh "\s-1ENTRY\s0"
.IX Subsection "\s-1ENTRY\s0"
<*pchunk> must be a \*(L"clean\*(R" chunk, returned by either
\fIRIFFIOChunkDescend()\fR or \fIRIFFIOChunkFinalize()\fR.

.rn }` ''


niff-1.02/niffsdk1.02/riffio/src/lib/riffio/chunks.c
#ifndef lint
static char rcsid[] = 
"$Id: chunks.c,v 1.2 1996/05/31 06:02:15 tim Exp $";
#endif
/*
 * Public Domain 1995,1996 Timothy Butler
 *
 * THIS DOCUMENT IS PROVIDED "AS IS" AND WITHOUT ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, WITHOUT LIMITATION, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE.
 *
 */
/***************************************************************************/
/*
 * NAME
 * ====
 * chunks - create, navigate, and operate on chunks 
 *
 * SYNOPSIS
 * ========
 * 
 * Chunk Creation
 * --------------
 * - RIFFIOChunkCreate()
 * - RIFFIOChunkFinalize()
 *
 * Chunk Navigation
 * ----------------
 * - RIFFIOChunkDescend()
 * - RIFFIOChunkAscend()
 * - RIFFIOChunkDataEnd()
 * - RIFFIOChunkEnd()
 * - RIFFIOChunkDataOffset()
 * - RIFFIOChunkSeekData()
 * 
 * Chunk Operations
 * ----------------
 * - RIFFIOChunkIsList()
 * 
 * DESCRIPTION
 * ===========
 * 
 * 
 */
/***************************************************************************/

#include <assert.h>
#include "riffio.h"

#ifndef NDEBUG
/*
 * The following variables and routines support a debugging feature 
 * to catch unbalanced calls to RIFFIOChunkCreate() and RIFFIOChunkFinalize()
 */
#include <stdlib.h>
static int _isCreateCalled = 0;            /* true if RIFFIOChunkCreate has
                                              been called at least once */
static int _CreateFinalizeMatch = 0;       /* counts outstanding calls of
                                              RIFFIOChunkCreate vs Finalize*/
static void _CheckCreateFinalize(void);

/*
 * _CheckCreateFinalize
 * ====================
 * Make sure RIFFIOChunkCreate() and RIFFIOChunkFinalize() are called
 * and equal number of times.
 *
 * This function will be registered with atexit the first time
 * RIFFIOChunkCreate is called.
 */
static void
_CheckCreateFinalize(void)
{
    assert(_CreateFinalizeMatch == 0);
}
#endif

/***************************************************************************/
/*
 * RIFFIOChunkCreate
 * =================
 * Start a new chunk in a RIFF file.
 */
RIFFIOSuccess
RIFFIOChunkCreate(RIFFIOFile *prf, RIFFIOChunk *pchunk) 
/*   
 * ENTRY
 * -----
 * - pchunk->fccId provides the new chunk id.
 *
 * - pchunk->fccType must contain a type id for Lists and Forms and
 *   is ignored otherwise.
 *
 * - pchunk->sizeData does not need to be correct.
 *
 * EXIT
 * ----
 * - Writes a chunk header to *prf according to *pchunk.
 *   RIFFIOChunkCreate() will write pchunk->sizeData to the file.
 * 
 * - Leaves *prf positioned at the start of the new chunk's data,
 * 
 * - If *pchunk is a form, then *prf is marked with the proper
 *   byte order. 
 * 
 * OBLIGATIONS
 *  ----------- 
 * - After writing the chunk contents, *pchunk
 *   must be finished with RIFFIOChunkFinalize().
 *  
 * - Don't change the <*pchunk> data members before calling
 *   RIFFIOChunkFinalize().  
 *
 * - If the pchunk->sizeData is correct when the chunk is finalized,
 *   then RIFFIOChunkFinalize() will not update the chunk's size in
 *   the file.
 *
 * ERRORS
 * ------
 *  On failure, the file position is undefined.  
 */
/***************************************************************************/
{

    char strModule[] = "RIFFIOChunkCreate";
    
    /* Check the arguments */
    assert(prf != 0);
    assert(pchunk != 0);
    
#ifndef NDEBUG
    /*
     * Is this the first time RIFFIOChunkCreate was called?
     */
    if (! _isCreateCalled)
    {
        /* yes, initialize the create/finalize balance debugging */
        assert(atexit(_CheckCreateFinalize) == 0);
        
        _isCreateCalled = 1;
    }
    
    /* 
     * Increment the number of times RIFFIOChunkCreate was
     * called relative to RIFFIOChunkFinalize
     */
    _CreateFinalizeMatch++;

#endif

    /*
     * Write the chunk id 
     */
    if (! RIFFIOWriteFOURCC(prf, pchunk->fccId))
    {
        RIFFIOError(strModule, "Failed to write chunk id");
        return RIFFIO_FAIL;
    }

    /*
     * If this is a form chunk, then mark the file with the specific
     * form type (byte order).
     * This has to be done before we write the size.
     *
     */
    switch(pchunk->fccId)
    {
      case RIFFIO_FOURCC_RIFF:
          prf->formType = RIFFIO_FORM_RIFF;
                
          break;
                
      case RIFFIO_FOURCC_RIFX:
          prf->formType = RIFFIO_FORM_RIFX;
          break;
    }

    /*
     * Write the chunk size
     */
    if (!RIFFIOWrite32(prf, pchunk->sizeData))
    {
        RIFFIOError(strModule, "Failed to write chunk size");
        return RIFFIO_FAIL;
    }

    /*
     * Remember the offset to the chunk's contents
     * This includes a form or list's FOURCC type.
     */
    pchunk->offsetData = RIFFIOTell(prf);


    /*
     * Write a list chunk's type
     */
    if (RIFFIOChunkIsList(pchunk))
    {
        if (!RIFFIOWriteFOURCC(prf, pchunk->fccType))
        {
            RIFFIOError(strModule, "Failed to write chunk type");
            return RIFFIO_FAIL;
        }
    }

    /*
     * Mark the chunk as dirty
     */
    pchunk->isDirty = 1;

    /*
     * Leave the file positioned at the start of either
     * the chunk's data or a list's first subchunk
     */
    return RIFFIO_OK;

}

/***************************************************************************/
/*
 * RIFFIOChunkFinalize
 * ===================
 * Finish writing a new chunk in a RIFF file.
 */ 
RIFFIOSuccess
RIFFIOChunkFinalize(RIFFIOFile *prf, RIFFIOChunk *pchunk)
/*
 * ENTRY
 * -----
 * <*pchunk> was returned by RIFFIOChunkCreate
 *
 * EXIT
 * ----
 * - Updates the chunk size in the file and in <*pchunk> (if necessary).
 *
 * - Writes a NUL pad byte (if necessary).
 *
 * - Leaves the file positioned after the chunk.
 *
 * ERRORS
 * ------
 * On failure, the file position is undefined.
 */
/***************************************************************************/
{

    char strModule[] = "RIFFIOChunkFinalize";

    RIFFIOSize   sizeNew;     /* Size of a "dirty" chunk */
    RIFFIOOffset offChunkEnd; /* Offset of the end of a "dirty" chunk
                               * not including any nul pad byte 
                               */


    /* Check the arguments */
    assert (prf != 0);
    assert (pchunk !=0);

    /* 
     * We should only be finalizing newly created chunks
     */
    assert (pchunk->isDirty);

#ifndef NDEBUG
        
    /* 
     * Decrement the number of times RIFFIOChunkCreate was called
     * relative to RIFFIOChunkFinalize
     */
    _CreateFinalizeMatch--;

#endif

    /*
     * We are at the end of chunk data
     * by definition.
     */
    offChunkEnd = RIFFIOTell(prf);
        
    /* 
     * Make sure we aren't trying to update a chunk with
     * its end before its beginnning.
     */
    assert( 
           (RIFFIOChunkIsList(pchunk) 
            && (offChunkEnd >= pchunk->offsetData+4))
           || 
           (!RIFFIOChunkIsList(pchunk) 
            && (offChunkEnd >= pchunk->offsetData)));

    /*
     * Calculate the size of the chunk's contents
     * not including any pad byte
     */
    sizeNew = offChunkEnd - (pchunk->offsetData);
        
    /*
     * Add a nul pad byte if the chunks contents is 
     * an odd number of bytes.
     * We don't count this in the chunk's size.
     */
    if (sizeNew % 2) 
    {
        if(! RIFFIOWrite8(prf, '\0'))
        {
            RIFFIOError(strModule, 
                        "Failed to write pad byte at end of chunk");
            return RIFFIO_FAIL;
        }
    }
        
    /* 
     * If the newly calculated chunk size differs
     * from its created value, then update the size
     */
    if (sizeNew != pchunk->sizeData)
    {
        /* Seek to the size field */
        if (! RIFFIOSeek(prf, pchunk->offsetData - 4, RIFFIO_SEEK_SET))
        {
            RIFFIOError(strModule, "Failed seek to update chunk size");
            return RIFFIO_FAIL;
        }
                        
        /* Re-write the size field */
        if (! RIFFIOWrite32(prf, sizeNew))
        {
            RIFFIOError(strModule, "Failed to write updated chunk size");
            return RIFFIO_FAIL;
        }
                        
        pchunk->sizeData = sizeNew;
    }

    /* 
     * Mark the chunk as having the correct size 
     */
    pchunk->isDirty = 0;
        
    /*
     * Leave the file positioned after the newly created chunk
     */
    if (! RIFFIOChunkAscend(prf, pchunk))
    {
        RIFFIOError(strModule, "Failed to ascend past new chunk");
        return RIFFIO_FAIL;
    }

    return RIFFIO_OK;

}


/***************************************************************************/
/*
 * RIFFIOChunkDescend
 * ==================
 * Read a chunk header from a RIFF file.
 */
RIFFIOSuccess
RIFFIOChunkDescend(RIFFIOFile *prf, RIFFIOChunk *pchunk)
/*
 * ENTRY
 * -----
 * <*prf> must be positioned at the start of a chunk.
 *
 * EXIT
 * ----
 * - Leaves the file positioned at the start of the chunk's contents 
 *   (either the first subchunk of a LIST or a normal chunk's data).
 *
 * - If the chunk is a RIFF or RIFX form, then
 *   marks the RIFFIOFile byte order accordingly.
 * 
 * - <*pchunk> holds the id, size, and possibly the
 *   list type of the newly read chunk.
 * 
 * ERRORS 
 * ------ 
 * On failure the file position is undefined.
 */ 
/***************************************************************************/
{

    char strModule[] = "RIFFIOChunkDescend";
        
    /* Check the arguments */
    assert(prf != 0);
    assert(pchunk != 0);

    /*
     * We are positioned at the start of a chunk.
     */

    /*
     * Read the chunk's id
     */
    if (!RIFFIOReadFOURCC(prf, &(pchunk->fccId)))
    {
        RIFFIOError(strModule, "Failed to read chunk id");
        return RIFFIO_FAIL;
    }

    /*
     * If the chunk id is RIFF or RIFX then mark the
     * file form type accordingly.
     * This must be done before we read the chunk size.
     */
    switch (pchunk->fccId)
    {
      case RIFFIO_FOURCC_RIFF:
        prf->formType = RIFFIO_FORM_RIFF;
        break;

      case RIFFIO_FOURCC_RIFX:
        prf->formType = RIFFIO_FORM_RIFX;
        break;
    }

    /*
     * Read the chunk size
     */
    if (! RIFFIORead32(prf, &(pchunk->sizeData)))
    {
        RIFFIOError(strModule, "Failed to read chunk size");
        return RIFFIO_FAIL;
    }

    /* 
     * Remember the offset of the chunk's contents
     * (includes a form or list's type)
     */ 
    pchunk->offsetData = RIFFIOTell(prf);

    /*
     * If the chunk is a form or list, 
     * then read its type
     */
    if (RIFFIOChunkIsList(pchunk))
    {
        if(pchunk->sizeData < 4)
        {
            RIFFIOError(strModule, 
                        "Found list data size less than four bytes");
            return RIFFIO_FAIL;
        }

        if(! RIFFIOReadFOURCC(prf, &(pchunk->fccType)))
        {
            RIFFIOError(strModule, "Failed to read chunk type");
            return RIFFIO_FAIL;
        }
    }

    /* 
     * Mark the chunk as "clean"
     */
    pchunk->isDirty = 0;

    /* 
     * Leave the user positioned at the beginning of the chunk's
     * data (not including list or form's type).
     */

    return RIFFIO_OK;

}

/***************************************************************************/
/*
 * RIFFIOChunkAscend 
 * =================
 * Advance a RIFF file past a chunk.
 */
RIFFIOSuccess
RIFFIOChunkAscend(RIFFIOFile *prf, const RIFFIOChunk *pchunk)
/*
 * ENTRY
 * -----
 * <*pchunk> was returned by either RIFFIOChunkDescend or
 * RIFFIOChunkFinalize.
 *
 * EXIT
 * ----
 * <*prf> is positioned at the end of <*pchunk>
 *
 * ERRORS
 * ------ 
 * On failure, the file position is undefined.
 * 
 */ 
/***************************************************************************/
{

    char strModule[] = "RIFFIOChunkAscend";


    /* Check the arguments */
    assert(prf != 0);
    assert(pchunk != 0);

        
    /* 
     * The chunk must have the correct size
     */
    assert (pchunk->isDirty == 0);

    /*
     * Leave the user positioned after the chunk and any
     * nul pad byte
     */
    if (! RIFFIOSeek(prf, 
                     pchunk->offsetData 
                     + pchunk->sizeData 
                     + (pchunk->sizeData % 2), 
                     RIFFIO_SEEK_SET))
    {
        RIFFIOError(strModule, "Failed to seek past chunk");
        return RIFFIO_FAIL; 
    }

    return RIFFIO_OK;

}



/***************************************************************************/
/*
 * RIFFIOChunkDataOffset
 * =====================
 * Return the file position of a chunk's data (or list's first subchunk)
 */
RIFFIOOffset
RIFFIOChunkDataOffset(const RIFFIOChunk *pchunk)
/***************************************************************************/
{
    assert (pchunk != 0);

    /*
     * Figure out where the chunk's contents start.
     * If it's a list then we have to account for the
     * type field
     */
    if (RIFFIOChunkIsList(pchunk))
        return pchunk->offsetData + 4;
    else
        return pchunk->offsetData;

}


/***************************************************************************/
/*
 * RIFFIOChunkSeekData
 * ===================
 * Seek to the start of a chunk's data or a list's contents
 */
RIFFIOSuccess
RIFFIOChunkSeekData(RIFFIOFile *prf, const RIFFIOChunk *pchunk)
/*
 * Note that this is different than RIFFIOChunkDescend() because
 * RIFFIOChunkSeekData() can be called regardless of the current 
 * file postion.
 * 
 * ENTRY 
 * ----- 
 * <*prf> may be positioned anywhere.
 *
 * EXIT 
 * ----
 * <*prf> is postioned at the start of a chunk's data or the first
 * subchunk of a list.
 * 
 * ERRORS 
 * ------
 * On error, the file postion is undefined.
 */
/***************************************************************************/
{
    const char   strModule[] = "RIFFIOChunkSeekData";

    assert (prf != 0);
    assert (pchunk != 0);

        
    if (! RIFFIOSeek(prf, 
                     RIFFIOChunkDataOffset(pchunk),
                     RIFFIO_SEEK_SET))
    {
        RIFFIOError(strModule, "Failed to seek to chunk contents");
        return RIFFIO_FAIL;
    }
        
    return RIFFIO_OK;
   
}

/***************************************************************************/
/*
 * RIFFIOChunkIsList
 * =================
 * Return true if a chunk is a form or a list.
 */
int
RIFFIOChunkIsList(const RIFFIOChunk *pchunk)
/***************************************************************************/
{
    assert(pchunk != 0);

    switch (pchunk->fccId)
    {
    case RIFFIO_FOURCC_RIFF:
    case RIFFIO_FOURCC_RIFX:
    case RIFFIO_FOURCC_LIST:
        return 1;
    default:
        return 0;
                
    }
}

/***************************************************************************/
/*
 * RIFFIOChunkDataEnd
 * ==================
 * Return true if a file is positioned at the end of a chunk's data.
 */
int
RIFFIOChunkDataEnd(RIFFIOFile *prf, const RIFFIOChunk *pchunk)
/*
 * The chunk data's end may be before any NUL pad byte.
 * 
 * ENTRY
 * -----
 * <*pchunk> must be a "clean" chunk, returned by either 
 * RIFFIOChunkDescend() or RIFFIOChunkFinalize().
 */
/***************************************************************************/
{

    RIFFIOOffset offFile; /* current file position */
        
    assert(prf != 0);
    assert(pchunk != 0);

    /*
     * We can't test "dirty" chunks
     */
    assert(pchunk->isDirty == 0);

    offFile = RIFFIOTell(prf);
        
    if (offFile == pchunk->offsetData + pchunk->sizeData) 
        return 1;
    else
        return 0;
        
}


/***************************************************************************/
/*
 * RIFFIOChunkEnd
 * ==============
 * Return true if a RIFF file is positioned at the end of a chunk.
 */
int
RIFFIOChunkEnd(RIFFIOFile *prf, const RIFFIOChunk *pchunk)
/* 
 * The chunk end is after any NUL pad byte.
 *
 * ENTRY
 * -----
 * <*pchunk> must be a "clean" chunk, returned by either 
 * RIFFIOChunkDescend() or RIFFIOChunkFinalize().
 */
/***************************************************************************/
{

    RIFFIOOffset offFile; /* File position */
        
    assert(prf != 0);
    assert(pchunk != 0);

    /*
     * We can't test "dirty" chunks
     */
    assert(pchunk->isDirty == 0);

    offFile = RIFFIOTell(prf);
        
    if (offFile == 
        pchunk->offsetData 
        + pchunk->sizeData 
        + (pchunk->sizeData % 2))
        return 1;
    else
        return 0;
        
}

